### PR TITLE
Handle the use of the AddressPrefixes field

### PR DESCRIPTION
**Pull request by ArielLima**
### Issue this PR addresses:

Currently if  **AddressPrefixes** is used **AddressPrefix** will have a value of nil which will cause a cluster to fail provisioning.

### What this PR does / why we need it:

If  **AddressPrefix**  is nil we try to validate  **AddressPrefixes**  instead. 

### Test plan for issue:

I altered one of the existing tests to use the  **AddressPrefixes** field, which I think covers the new and old use cases

### Is there any documentation that needs to be updated for this PR?

I am not sure so if there is please point me to it.

### CODE REVIEW

**Comment by ArielLima** [comment]
[Work Item](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/12362612/)

File: .gitignore [hunk 1]
```diff
@@ -27,3 +27,4 @@ gomock_reflect_*
 /deploy/config.yaml
 **/*.swp
 /portal/node_modules/
+.idea*
```
Comment (petrkotas) [inline .gitignore:30]: I am happy to ack this change, I have but a simple ask. Please make it a separate commit so it is not mixed with the code changes. Thank you.

File: pkg/api/validate/dynamic/dynamic_test.go [hunk 2]
```diff
@@ -436,7 +436,7 @@ func TestValidateCIDRRanges(t *testing.T) {
 						{
 							ID: &workerSubnet,
 							SubnetPropertiesFormat: &mgmtnetwork.SubnetPropertiesFormat{
-								AddressPrefix: to.StringPtr("10.0.1.0/24"),
+								AddressPrefixes: to.StringSlicePtr([]string{"10.0.1.0/24"}),
```
Comment (petrkotas) [inline pkg/api/validate/dynamic/dynamic_test.go:439]: The logic of the flow changed, therefore the test need to change too. Would you please modify the test to cover both logical flows? I understand it is a big change, so my appologies. We need to watch for regressions in our codebase.

File: pkg/api/validate/dynamic/dynamic.go [hunk 3]
```diff
@@ -287,11 +287,24 @@ func (dv *dynamic) validateCIDRRanges(ctx context.Context, subnets []Subnet, add
 			return err
 		}
 
-		_, net, err := net.ParseCIDR(*s.AddressPrefix)
-		if err != nil {
-			return err
+		// Since AddressPrefix seems to be deprecated for addressPrefixes we need this to validate the CIDR of the new
```
Comment (petrkotas) [inline pkg/api/validate/dynamic/dynamic.go:290]: Thank you for clearly documenting the intention and marking the TODO. I highly appreciate the clarity here.

**Review by petrkotas — DISMISSED** [review]
Hi @ArielLima thank you for covering this. I like the simplicity of the solution. I found we small items I would love you to consider as an improvement over your already good solution. After that, we are good to go.

**Review by bennerv — DISMISSED** [review]
lgtm

File: pkg/api/validate/dynamic/dynamic_test.go [hunk 4]
```diff
@@ -450,23 +478,25 @@ func TestValidateCIDRRanges(t *testing.T) {
 				tt.modifyOC(oc)
 			}
 
-			vnetClient := mock_network.NewMockVirtualNetworksClient(controller)
-			if tt.vnetMocks != nil {
-				tt.vnetMocks(vnetClient, vnet)
-			}
+			for _, vnet := range vnets {
```
Comment (m1kola) [inline pkg/api/validate/dynamic/dynamic_test.go:481]: I see that we loop over few flavours of vnet definition, but I think failure in this loop will be hard to debug.

I would add another another `t.Run` somewhere here and make sure to give a good name to it.

Something like:

```go
vnet := map[string]mgmtnetwork.VirtualNetwork{
	"vnet with AddressPrefix": {
		ID:       &vnetID,
		// ...
	},
	"vnet with AddressPrefixes": {
		ID:       &vnetID,
		// ...
	},
}

for vnetCase, vnet := range vnets {
	t.Run(vnetCase, func(t *testing.T) {
		// ...
	})
}
```

File: pkg/api/validate/dynamic/dynamic.go [hunk 5]
```diff
@@ -287,11 +287,24 @@ func (dv *dynamic) validateCIDRRanges(ctx context.Context, subnets []Subnet, add
...
+		// field AddressPrefixes
+		// TODO: In theory we should be able to only check AddressPrefixes in the future
```
Comment (m1kola) [inline pkg/api/validate/dynamic/dynamic.go:292]: Can we find out now? So we don't have to have this to do?

**Review by m1kola — COMMENTED** [review]
A small suggestion about tests and a question. Other than that - looks good.

File: pkg/api/validate/dynamic/dynamic.go (hunk 5 above)
Comment (ArielLima) [reply pkg/api/validate/dynamic/dynamic.go:292]: @m1kola I was told by @bennerv not to make that change yet either way because of dependencies? @bennerv could you give some more insight?

File: pkg/api/validate/dynamic/dynamic_test.go (hunk 4 above)
Comment (ArielLima) [reply pkg/api/validate/dynamic/dynamic_test.go:483]: @m1kola Thank you for the input, I added names to the vnets and log those with any error. This should make debugging straight forward + it doesn't require restructuring the whole test.

File: pkg/api/validate/dynamic/dynamic.go (hunk 5 above)
Comment (bennerv) [reply pkg/api/validate/dynamic/dynamic.go:292]: @m1kola @ArielLima 

`addressPrefixes` was moved from its own `AddressSpace` property to a top-level array property in the 2018-08-01 Network Resource provider release.  I don't think they're deprecating `addressPrefix` (singular) so we have to account for both. 

If you create a vnet / subnet address space using the current azure-cli, it will populate `addressPrefixes` (plural), but `addressPrefix` (singular) will be null.  

The `addressPrefixes` (plural) is used for dual-stack (ipv4/ipv6) networks, but the two are not mutually exclusive (you can have `addressPrefixes` specified without any ipv6 address space).  

I think this solution is the right way forward although it seems like a hack, but we have to account for the design of the NRP's API when we enumerate these subnets.

Comment (bennerv) [reply pkg/api/validate/dynamic/dynamic.go:292]: @ArielLima 
It's possible they'll deprecate the field in the future, but we will hit that problem when we bump the consumed API version in the azure-sdk-for-go.  For now it might make sense to just remove the comment about deprecation and explain that you can specify it in multiple places and both are valid configs according to the NRP.

**Comment by bennerv** [comment]
@ArielLima do you mind squashing the 4 commits into one or two commits here?  Will keep release history concise going forward as we will soon have release note integration based on commit message.

**Review by bennerv — APPROVED** [review]
LGTM - merge on green ci/e2e

**Comment by bennerv** [comment]
@petrkotas dismissing your review and merging as all is resolved from your previous comments.
